operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java

```diff
@@ -16,82 +16,162 @@
+import org.keycloak.operator.v2alpha1.crds.KeycloakStatusBuilder;
 
 import java.net.URL;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class KeycloakDeployment extends OperatorManagedResource {
+
+    public static final Pattern CONFIG_SECRET_PATTERN = Pattern.compile("^\\$\\{secret:([^:]+):(.+)}$");
+
+    private final Config config;
+    private final Keycloak keycloakCR;
+    private final Deployment existingDeployment;
+    private final Deployment defaultDeployment;
+    private final Set<String> configSecretsNames;
+
+    public KeycloakDeployment(KubernetesClient client, Config config, Keycloak keycloakCR, Deployment existingDeployment) {
+        super(client, keycloakCR);
+        this.config = config;
+        this.keycloakCR = keycloakCR;
+
+        if (existingDeployment != null) {
+            Log.info("Existing Deployment provided by controller");
+            this.existingDeployment = existingDeployment;
+        }
+        else {
+            Log.info("Trying to fetch existing Deployment from the API");
+            this.existingDeployment = fetchExistingDeployment();
+        }
+
+        // process the CR
+
+        URL url = this.getClass().getResource("/base-keycloak-deployment.yaml");
+        defaultDeployment = client.apps().deployments().load(url).get();
 
-import static org.keycloak.operator.v2alpha1.crds.KeycloakStatus.State.*;
+        defaultDeployment.getMetadata().setName(getName());
+        defaultDeployment.getMetadata().setNamespace(getNamespace());
+        defaultDeployment.getSpec().getSelector().setMatchLabels(Constants.DEFAULT_LABELS);
+        defaultDeployment.getSpec().setReplicas(keycloakCR.getSpec().getInstances());
+        defaultDeployment.getSpec().getTemplate().getMetadata().setLabels(Constants.DEFAULT_LABELS);
 
-public class KeycloakDeployment {
+        Container container = defaultDeployment.getSpec().getTemplate().getSpec().getContainers().get(0);
+        container.setImage(Optional.ofNullable(keycloakCR.getSpec().getImage()).orElse(config.operand().image()));
 
-    KubernetesClient client = null;
+        var distConfig = new HashMap<>(Constants.DEFAULT_DIST_CONFIG);
+        if (keycloakCR.getSpec().getDistConfig() != null) {
+            distConfig.putAll(keycloakCR.getSpec().getDistConfig());
+        }
 
-    KeycloakDeployment(KubernetesClient client) {
-        this.client = client;
+        Set<String> configSecretsNames = new HashSet<>();
+        List<EnvVar> configEnvVars = distConfig.entrySet().stream()
+                .map(e -> {
+                    EnvVarBuilder builder = new EnvVarBuilder().withName(e.getKey());
+                    Matcher matcher = CONFIG_SECRET_PATTERN.matcher(e.getValue());
```

Comment:
ok, I think I got how you want this to work @vmuzikar , can you elaborate on the reasoning for spinning out a custom "vault" integration on top of Kubernetes when the same functionality is offered by `envFrom`?

Reply:
The motivation behind this is that I did not want to tie the dist config to env vars, at least not from the API PoV that we provide in the CRD. What if we decide in the future that we want to implement it differently and use e.g. config files and not env vars? I was even wondering if we should use some "translation" from Quarkus to allow users to use the config keys as in config files or system properties (the dot notation) as I see it more generic and not the env vars notation (capitals and snake case).

Reply:
I was thinking about this indirection and I do believe that it brings a number of downsides:

- not standard
- it works as a re-name that can be misleading
- the re-naming makes it harder to debug issues
- it's not operator business to mangle Keycloak distribution properties
- custom implementation of an already existing feature
- no proper feedback for the user in case of misconfiguration
- subject to subtle and nasty bugs (e.g. add a space `${ source:`)
- additional logic to be handled

I do see a single interesting use-case that, potentially(not with the current implementation), this feature can bring on the table that is:

 - fetch env variable values from external Secrets (such as the credentials of the DB from the Secret created by the Crunchy operator)
 
 to get there we would need to add support for namespaces and Json parsing.
 
All in all, this PR is intended to be a `Baseline` for the deployment, I believe we agree that this is already a `feature`, so, my suggestion is to leave the passwords in clear text for now, remove this logic, and open a issue to re-visit/re-iterate/better discuss this feature in a separate PR.